### ISSUE TITLE
Use a bootstrap environment to run anaconda-project's main logic

### ISSUE THREAD

**Issue by havocp** [Anaconda-Platform/anaconda-project#86]
## Problem to be solved

There's an issue that a project may want to control the version of anaconda-project used with that project, the version of Python used to launch anaconda-project, and in the future when we have plugins the versions of the plugins.

To accomplish this, we could have a "bootstrap environment." The `anaconda-project` executable would be as thin as possible; its role would be to check that the bootstrap environment exists and is up-to-date, and then to re-exec the "real" anaconda-project with the bootstrap environment activated.

With this setup, the only part of anaconda-project.yml/anaconda-project-lock.yml that has to stay compatible indefinitely would be the part that affects the bootstrap environment. The "outer" anaconda-project executable would ignore all other configuration.

## How it would work, roughly

When you run `anaconda-project foo something`, the following would happen:
 
 * Each anaconda-project.yml would define a “bootstrap” env spec. The naming could be discussed, but perhaps it’s in the env_specs section and always named `anaconda_project` or `anaconda_project_bootstrap` or similar.
 * By default the bootstrap env spec would contain just the `anaconda-project` package, but if you use plugins you’d include their packages in here too
 * The bootstrap env spec can be locked in anaconda-project-lock.yml the same as any other env spec
 * The bootstrap env would be created in PROJECT_DIR/envs the same as any other
 * On startup, anaconda-project will check whether it is the anaconda-project located inside the bootstrap environment for the project you’ve run it on.
 * If we are in the current project’s bootstrap environment, we continue as we do today
 * If we are not in the current project’s bootstrap environment, we create that bootstrap environment, activate it, and then re-run our argv.

So projects’ bootstrap environment will be created by some arbitrary anaconda-project version, but they can assume all other anaconda-project activities will be done by the version they have specified in their bootstrap env spec.
 
## What about using anaconda-project via an API

Right now some GUI tools use the anaconda-project API rather than the anaconda-project executable. These would be calling the API from their own environment and not the bootstrap environment, which would break (especially once we have plugins).

To solve this, we could:
 * run these tools in the bootstrap environment, OR
 * add some sort of out-of-process JSON/REST mode (start an "anaconda project server", essentially) and have GUI tools talk to that, instead of using an in-process API

The not-very-useful api.py wrapper we have now could maybe be dropped, and replaced with an API that transparently starts/stops and talks to an external process.

This could be the bulk of the work in implementing the bootstrap environment approach.

## Other implementation details
 
 * We might introduce a special “bootstrap mode” of the project validation code in anaconda_project/project.py, which would ignore everything except the bootstrap env spec, since we don’t want to choke on something the “child” anaconda-project version (or its plugins) will be able to understand. Alternatively, maybe we just ignore all validation errors.
 * Should the “bootstrap” env spec inherit global `packages:`/`channels:`/etc. just like the other env specs? Doing this means the bootstrap env would be a lot bigger than it has to be in most cases (by default it would contain `anaconda` metapackage). If we don’t do this, maybe the bootstrap env spec should not be in the env_specs section but should be pulled out into its own toplevel section, to make clear that it’s special? Both options here seem a little unfortunate.
 * Should the child anaconda-project check for up-to-dateness of its own bootstrap environment? Example of when this would have mattered: when we added version locking, if you had an old anaconda-project on the outside, it wouldn’t have locked the bootstrapped env, but the child anaconda-project would know to set package versions to the locked ones.
 * there’s a pathology possible where outer and inner anaconda-project “fight” over environment contents: every time you launch anaconda-project, the outer one might change some packages, then the inner one change them, then outer change them back again on next invocation, then inner changes them again, etc.
 * A middle ground could be that the outer a-p creates a bootstrap env containing ONLY python and anaconda-project itself, then starts the inner which makes any other bootstrap env changes, and then the inner restarts itself if it modified the bootstrap env? too complex?
 * If the child anaconda-project modifies installs new versions of itself or its plugins, it has to restart itself.
 * With the bootstrap setup, we can make it a hard error if we see unknown fields in anaconda-project.yml while in the “child” anaconda-project, because there’s no reason to be “forwards compatible”; if you need anaconda-project 2.0, include that in your bootstrap env spec.
 * We could include an error-with-autofix as follows:
    - If there’s an unknown field, it’s an error
    - The autofix will suggest installing `anaconda-project-plugin-<fieldname>` in the bootstrap env spec. So if you have a `jupyter:` field in your command and don’t put anaconda-project-plugin-jupyter in your bootstrap env spec, we offer to do that for you.
    - As with all autofixes, we auto-apply them when you `anaconda-project init`, keeping the goal in mind that `anaconda-project init` on a directory containing only a notebook file will result in a working project.
 * The default bootstrap environment might be copied from the “outer” anaconda-project (same anaconda-project version, same plugins)?
 * We could additionally short-circuit and avoid creating the bootstrap environment if the outer root environment “meets” the bootstrap env spec. This could save a lot of time/disk in the common case, but also creates a different codepath that could deviate from what will happen on someone else’s machine.
 * A subtlety of this is that anaconda-project should not load plugins that aren’t listed in the bootstrap env spec, even if those plugins are present in the environment. This is because the “outer” environment could contain a superset of the plugins we actually want. So the optimization gets complicated.
 * The behavior of `anaconda-project init` will now depend on the plugins you have installed. This could get ugly. (If you try to `init` a directory containing a notebook file, it would work if you’ve installed the jupyter plugin but not otherwise, for example.) We would need to install the common plugins in our default env we init from, perhaps.

## Bootstrap all of Anaconda
 
Future stretch goal: standalone binary instead of downloading/activating miniconda.

There’s a world where you could download a single file “anaconda-project” and never install a root environment anywhere, only bootstrap project-specific environments. One way to implement this would be a shell script or other standalone script called “anaconda-project” that creates the bootstrap environment by downloading miniconda, installing it to envs/anaconda_project, and then running child anaconda-project which would be the actual anaconda-project and add stuff to the bootstrap environment / make it conform to the bootstrap env spec. You could even check this anaconda-project script in to your project's git repo, making the project entirely standalone.

**Comment by havocp** [comment]
cc @fpliger @bollwyvl

**Comment by fpliger** [comment]
Cool, good write up!

On quick thought about it is that it shouldn't be a show stopper for adding plugins first.... right? :)

**Comment by havocp** [comment]
Agreed, plugin mechanism is separate. However without the bootstrap feature to use a project that needs a plugin you would have to manually install the plugin before using anaconda-project.

**Comment by mcg1969** [comment]
Thanks for opening this issue, and apologies for the long delay in responding.

Anaconda Project has not been actively developed in some time and has largely been superseded by newer alternatives:

* [pixi](https://pixi.sh/) and [uv](https://docs.astral.sh/uv/) represent the first truly successful successors.
* [conda-workspaces](https://github.com/conda-incubator/conda-workspaces) is a promising pixi-compatible alternative for users who prefer to stay closer to the conda ecosystem.

Going forward, work on this repository is focused on critical bug fixes and on tooling to help users migrate existing Anaconda Project projects to pixi (and, by extension, conda-workspaces). See the [Pixi support](https://anaconda-project.readthedocs.io/en/latest/pixi-support.html) page in our documentation for details on the `anaconda-project export-pixi` command, which can convert an existing `anaconda-project.yml` into a `pixi.toml` for you.

The `anaconda-project` package itself will remain available, so existing projects relying on its stable feature set can continue to be supported. However, we do not anticipate further enhancement work, and for that reason we are closing out the bulk of our open issue backlog — including this one.

If this issue describes a critical regression in the stable feature set (as opposed to an enhancement or an edge case in a feature we no longer plan to expand) and you have a minimal reproducer against a current release, please feel free to reopen and we will take another look.

Thank you for your interest in Anaconda Project.
